# Tarea 03 — Procesamiento de Datos con Pandas

**Instrucciones:** Completa cada celda de código con tu respuesta.  
Trabajarás con el dataset del Titanic para practicar transformaciones reales.

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


---
## Parte 1 — Exploración y selección

**P1.** Imprime los nombres de todas las columnas y el número total de filas y columnas del DataFrame.

In [2]:
print(df.columns)
print(df.shape)


Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')
(891, 12)


**P2.** Usa `.loc` para seleccionar las filas 10 a 15 (inclusive) y solo las columnas `Name`, `Age` y `Survived`.

In [3]:
print(df.loc[10:15, ["Name", "Age", "Survived"]])


                                    Name   Age  Survived
10       Sandstrom, Miss. Marguerite Rut   4.0         1
11              Bonnell, Miss. Elizabeth  58.0         1
12        Saundercock, Mr. William Henry  20.0         0
13           Andersson, Mr. Anders Johan  39.0         0
14  Vestrom, Miss. Hulda Amanda Adolfina  14.0         0
15      Hewlett, Mrs. (Mary D Kingcome)   55.0         1


---
## Parte 2 — Valores nulos y limpieza

**P3.** Cuenta cuántos valores nulos hay en cada columna. ¿Cuáles columnas tienen nulos?

In [4]:
print(df.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


**Respuesta:** Las columnas con valores nulos son: Age, Cabin y Embarked

**P4.** Rellena los valores nulos de la columna `Age` con la **mediana** de esa columna.  
Verifica que ya no haya nulos en `Age`.

In [6]:
#Rellenar nulos de Age
mediana = df["Age"].median()
df["Age"] = df["Age"].fillna(mediana)

#Verificar que no haya nulos
print("Nulos en Age:", df["Age"].isnull().sum())

Nulos en Age: 0


---
## Parte 3 — Nuevas columnas y agrupaciones

**P5.** Crea una nueva columna llamada `FamilySize` que sea la suma de `SibSp` + `Parch` + 1 (el pasajero mismo).

In [9]:
df["FamilySize"] = df["SibSp"] + df["Parch"] + 1
print(df["FamilySize"].head())

#Columnas juntas para comparar
print(df[["SibSp", "Parch", "FamilySize"]].head(10))

0    2
1    2
2    1
3    2
4    1
Name: FamilySize, dtype: int64
   SibSp  Parch  FamilySize
0      1      0           2
1      1      0           2
2      0      0           1
3      1      0           2
4      0      0           1
5      0      0           1
6      0      0           1
7      3      1           5
8      0      2           3
9      1      0           2


**P6.** Usando `groupby`, calcula la **tasa de supervivencia promedio** (columna `Survived`) por clase (`Pclass`).  
¿Qué clase tuvo mayor probabilidad de sobrevivir?

In [10]:
print(df.groupby("Pclass")["Survived"].mean())


Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64


**Respuesta:** La clase con mayor supervivencia fue: 1

**P7.** Calcula la edad promedio y la tarifa (`Fare`) promedio, agrupando por `Sex` y `Pclass`.  
Usa `.groupby()` con `.agg()`.

In [11]:
print(df.groupby(["Sex", "Pclass"]).agg({"Age": "mean", "Fare": "mean"}))

                     Age        Fare
Sex    Pclass                       
female 1       33.978723  106.125798
       2       28.703947   21.970121
       3       23.572917   16.118810
male   1       38.995246   67.226127
       2       30.512315   19.741782
       3       26.911873   12.661633


---
## Parte 4 — Ordenamiento y duplicados

**P8.** Ordena el DataFrame por `Fare` de mayor a menor y muestra las primeras 5 filas.  
¿Cuánto pagó el pasajero más caro?

In [12]:
print(df.sort_values("Fare", ascending=False).head())


     PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
88            89         1       1          Fortune, Miss. Mabel Helen   
27            28         0       1      Fortune, Mr. Charles Alexander   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
258  female  35.0      0      0  PC 17755  512.3292          NaN        C   
737    male  35.0      0      0  PC 17755  512.3292         B101        C   
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
88   female  23.0      3      2     19950  263.0000  C23 C25 C27        S   
27     male  19.0      3      2     19950  263.0000  C23 C25 C27        S   

     FamilySize  
258           1  
737           1  
679           2  
88            6  
27

#Pago el pasajero mas caro: $512.3292

**P9.** Verifica si hay filas duplicadas en el DataFrame. ¿Cuántas hay?

In [13]:
print("Filas duplicadas:", df.duplicated().sum())

Filas duplicadas: 0
